lsp-mode.el

```diff
@@ -4601,34 +4601,53 @@ Others: TRIGGER-CHARS"
                 (error)))
       item))
 
-(defun lsp--extract-line-from-buffer (pos)
-  "Return the line pointed to by POS (a Position object) in the current buffer."
-  (let* ((point (lsp--position-to-point pos))
-         (inhibit-field-text-motion t))
+(defun lsp--xref-make-items (filename ranges)
+  "Return a list of `xref-item's from a list of corresponding RANGES in FILENAME."
+  (let ((inhibit-field-text-motion t)
+        (ranges (-sort (lambda (range-a range-b)
+                         (-let* ((pos-start-a (gethash "start" range-a))
+                                 (line-num-a (gethash "line" pos-start-a))
+                                 (col-num-a (gethash "character" pos-start-a))
+                                 (pos-start-b (gethash "start" range-b))
+                                 (line-num-b (gethash "line" pos-start-b))
+                                 (col-num-b (gethash "character" pos-start-b)))
+                           (lsp--line-col-comparator (cons line-num-a col-num-a)
+                                                     (cons line-num-b col-num-b))))
+                       ranges))
+        (curr-line 0)
+        accum)
     (save-excursion
-      (goto-char point)
-      (buffer-substring-no-properties (line-beginning-position)
-                                      (line-end-position)))))
-
-(defun lsp--xref-make-item (filename range)
-  "Return a xref-item from a RANGE in FILENAME."
-  (let* ((pos-start (gethash "start" range))
-         (pos-end (gethash "end" range))
-         (line (lsp--extract-line-from-buffer pos-start))
-         (start (gethash "character" pos-start))
-         (end (gethash "character" pos-end))
-         (len (length line)))
-    (add-face-text-property (max (min start len) 0)
-                            (max (min end len) 0)
-                            'highlight t line)
-    ;; LINE is nil when FILENAME is not being current visited by any buffer.
-    (xref-make (or line filename)
-               (xref-make-file-location filename
-                                        (1+ (gethash "line" pos-start))
-                                        (gethash "character" pos-start)))))
+      (save-restriction
+        (widen)
+        (goto-char (point-min))
+        (dolist (range ranges)
```

Comment:
is there a reason for preferring `dolist` over `-map`?

Reply:
I prefer to use `-map` with projection functions that do not mutate anything outside the scope they define, and to use `dolist` for impure functions.